dds/DCPS/DataReaderImpl.cpp

```diff
@@ -3282,6 +3282,56 @@ DataReaderImpl::get_ice_endpoint()
   return TransportClient::get_ice_endpoint();
 }
 
+DDS::ReturnCode_t DataReaderImpl::setup_deserialization()
+{
+  const DDS::DataRepresentationIdSeq repIds =
+    get_effective_data_rep_qos(qos_.representation.value);
+  bool success = false;
+  if (cdr_encapsulation()) {
+    Encoding::Kind encoding_kind;
+    for (CORBA::ULong i = 0; i < repIds.length(); ++i) {
+      if (repr_to_encoding_kind(repIds[i], encoding_kind)) {
+        if (Encoding::KIND_XCDR1 == encoding_kind ||
+            Encoding::KIND_XCDR2 == encoding_kind) {
+          if (decoding_modes_.find(encoding_kind) == decoding_modes_.end()) {
+            Encoding encoding(encoding_kind, swap_bytes());
+            decoding_modes_.insert(std::make_pair(encoding_kind, encoding));
+            success = true;
+          }
+        } else if (::OpenDDS::DCPS::DCPS_debug_level >= 2) {
+          //Valid but incompatible data representation
+          ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) ")
+                     ACE_TEXT("DataReaderImpl::setup_deserialization: ")
+                     ACE_TEXT("Skip %C data representation\n"),
+                     Encoding::kind_to_string(encoding_kind).c_str()));
+        }
+      } else if (::OpenDDS::DCPS::DCPS_debug_level) {
+        // Invalid data representation
+        ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: ")
+                   ACE_TEXT("DataReaderImpl::setup_deserialization: ")
+                   ACE_TEXT("Encounter invalid data representation\n")));
+        return DDS::RETCODE_ERROR;
+      }
+    }
+  } else {
+    Encoding encoding(Encoding::KIND_UNALIGNED_CDR, swap_bytes());
+    decoding_modes_.insert(std::make_pair(Encoding::KIND_UNALIGNED_CDR, encoding));
```

Comment:
This still needs to be used in decoding.

Reply:
I added a few lines in `dds_demarshal` to check if the encoding kind of the received data sample also appears in `decoding_modes_`. Please let me know if that's all we need to do with `decoding_modes_` or if there is any other logic we need to do with it.

Reply:
`lookup_instance` will also need to be updated. In addition to that I think the `decoding_modes_` can just be a `std::set<Encoding::Kind>`, since the actual `Encoding` object isn't going to be used.

Reply:
`lookup_instance` is updated now.